java/vector/src/main/codegen/data/ValueVectorTypes.tdd

```diff
@@ -123,6 +123,21 @@
       ]
     },
     {
+      major: "Fixed",
+      width: -1,
```

Comment:
FixedSizeBinary has a variable width, I am not sure what integer value to put here. I went ahead and put -1 to mark it as a special case. This changes also leads to quite some changes of if statements in the codegen template FixedValueVectors.java.

Reply:
I am not sure about this change. FixValueVectors have static type width that is specified in ValueVectorTypes.tdd. Having type width to be dynamic would complicated the template and result in inconsistency. 

Is there a reason FixedSizeBinary needs to have dynamic type witdh? i.e., can we have byteWitdh = 1 and use the type width in the Schema to retrieve and store values? This would simplify things.

cc @wesm

Reply:
The bottom line is that FixedSizeBinaryVector class should have a way to access the information about the dynamic type width. The only way I found is by changing the static type width in the class to non-static and initialize in the constructor. Did I miss a better way to pass the information? I don't like to put a lot of if/elseif/else statements in a codegen template either :)

Regarding the value being set to a positive integer like 1, it is highly misleading for readers in my opinion.

Reply:
I see. Yeah byteWidth = 1 wouldn't work because of how the template works. But I think we should still have typeBitWidth = 8 in the type layout in the schema to match Binary type.

Reply:
@icexelloss, the VarBinary type has a width of 4 (instead of 8) which make sense because each element has a 4-byte offset. Do you think it is better to use 4 as well? Btw, VarBinary is using a different template.

Reply:
Offset is 4 bytes wide. Data width is 1 byte wide.

Reply:
I am talking about the type layout in the schema. Not the constant in the vector class. Sorry for the confusion.

Reply:
I think we should probably refactor the templates some before introducing this change. In reality, I'd like to propose that we rework the vector hiearchy entirely and work much more like ArrowBuf and it's related parents work. Basically, let's implement to classes:

BaseFixedVector (would take a dynamic size variable)
BaseVariableVector

These would both be non-code generated classes. Then we could generate code generated subclasses of these that only implement methods that need code generation for performance. Most of these methods would probably appear as custom methods in mutator/accessors.

Then, the changes for a new Custom fixed width vector would just be specializations of the BaseFixedWidthVector called something like FixedBinaryVector and FixedVarcharVector and in those specialization, the width in the mutators and accessors could come from a instance variable instead of a constant.

I believe this would allow us to simplify the codebase and make maintenance of the base classes much easier.

Reply:
I opened up https://issues.apache.org/jira/browse/ARROW-1463 to cover this.

Reply:
sounds good to me. I'll wait for the refactoring to complete and come back to modify this PR